File: src/services/evaluators/schemaGenerator.ts
import type { EvaluatorOutputField } from '@/types';

/**
 * Generate JSON schema from user-defined output fields
 * This schema is passed to LLM for structured output enforcement
 */
export function generateJsonSchema(fields: EvaluatorOutputField[]): object {
  const properties: Record<string, object> = {};
  const required: string[] = [];
  
  fields.forEach(field => {
    properties[field.key] = generateFieldSchema(field);
    required.push(field.key); // All fields are required
  });
  
  return {
    type: 'object',
    properties,
    required,
    additionalProperties: false, // Strict schema
  };
}

function generateFieldSchema(field: EvaluatorOutputField): object {
  const baseSchema: Record<string, unknown> = {
    description: field.description,
  };
  
  switch (field.type) {
    case 'number':
      return { ...baseSchema, type: 'number' };
    
    case 'text':
      return { ...baseSchema, type: 'string' };
    
    case 'boolean':
      return { ...baseSchema, type: 'boolean' };
    
    case 'array':
      return {
        ...baseSchema,
        type: 'array',
        items: generateArrayItemSchema(field),
      };
    
    default:
      return { ...baseSchema, type: 'string' };
  }
}

function generateArrayItemSchema(field: EvaluatorOutputField): object {
  // If no schema defined, default to string array
  if (!field.arrayItemSchema) {
    return { type: 'string' };
  }
  
  const { itemType, properties } = field.arrayItemSchema;
  
  // Simple types
  if (itemType === 'string') return { type: 'string' };
  if (itemType === 'number') return { type: 'number' };
  if (itemType === 'boolean') return { type: 'boolean' };
  
  // Object type - build properties schema
  if (itemType === 'object' && properties && properties.length > 0) {
    const objectProperties: Record<string, object> = {};
    const required: string[] = [];
    
    properties.forEach(prop => {
      objectProperties[prop.key] = {
        type: prop.type,
        description: prop.description,
      };
      required.push(prop.key);
    });
    
    return {
      type: 'object',
      properties: objectProperties,
      required,
    };
  }
  
  // Fallback to string
  return { type: 'string' };
}

File: backend/app/services/evaluators/schema_generator.py
"""JSON Schema generator from field-based output definitions.

Ported from src/services/evaluators/schemaGenerator.ts — converts
EvaluatorOutputField[] (visual builder format) to JSON Schema for
structured LLM output enforcement.
"""


def generate_json_schema(fields: list[dict]) -> dict:
    """Convert field-based output schema to JSON Schema.

    Args:
        fields: List of field definitions, each with:
            - key: str (field name)
            - type: str ("number", "text", "boolean", "array")
            - description: str
            - arrayItemSchema: dict | None (for array type)

    Returns:
        JSON Schema object:
        {
            "type": "object",
            "properties": {...},
            "required": [...],
            "additionalProperties": false
        }
    """
    properties = {}
    required = []

    for field in fields:
        properties[field["key"]] = _generate_field_schema(field)
        required.append(field["key"])

    return {
        "type": "object",
        "properties": properties,
        "required": required,
        "additionalProperties": False,
    }


def _generate_field_schema(field: dict) -> dict:
    """Generate JSON Schema for a single field."""
    base = {}
    if field.get("description"):
        base["description"] = field["description"]

    field_type = field.get("type", "text")

    if field_type == "number":
        return {**base, "type": "number"}
    elif field_type == "text":
        return {**base, "type": "string"}
    elif field_type == "boolean":
        return {**base, "type": "boolean"}
    elif field_type == "array":
        return {
            **base,
            "type": "array",
            "items": _generate_array_item_schema(field),
        }
    else:
        return {**base, "type": "string"}


def _generate_array_item_schema(field: dict) -> dict:
    """Generate JSON Schema for array items."""
    item_schema = field.get("arrayItemSchema")
    if not item_schema:
        return {"type": "string"}

    item_type = item_schema.get("itemType", "string")

    if item_type == "string":
        return {"type": "string"}
    if item_type == "number":
        return {"type": "number"}
    if item_type == "boolean":
        return {"type": "boolean"}

    if item_type == "object":
        props = item_schema.get("properties", [])
        if props:
            object_properties = {}
            obj_required = []
            for prop in props:
                object_properties[prop["key"]] = {
                    "type": prop.get("type", "string"),
                }
                if prop.get("description"):
                    object_properties[prop["key"]]["description"] = prop["description"]
                obj_required.append(prop["key"])
            return {
                "type": "object",
                "properties": object_properties,
                "required": obj_required,
            }

    return {"type": "string"}
